### PR TITLE
refactor: reduce allocations in push down filter

### PR DESCRIPTION
**Pull request by erratic-pattern**
Not sure which issue to link here, but here's some changes to reduce allocations in `PushDownFilter`, and also clean the code up a bit.

### CODE REVIEW

**Review by comphead — COMMENTED** [review]
thanks @erratic-pattern 
please help me to understand how the approach in the PR differs from what we have now. Its  replaces clone on Arcs to Arc::clone, which imho the same https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references ?

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -802,11 +802,11 @@ impl OptimizerRule for PushDownFilter {
                     replace_map.insert(expr.display_name()?, expr.clone());
                 }
                 let replaced_push_predicates = push_predicates
-                    .iter()
-                    .map(|expr| replace_cols_by_name(expr.clone(), &replace_map))
+                    .into_iter()
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:805]: this looks like one example of less cloning

File: datafusion/optimizer/src/push_down_filter.rs [hunk 2]
```diff
@@ -829,12 +829,12 @@ impl OptimizerRule for PushDownFilter {
             }
             LogicalPlan::Join(join) => push_down_join(join, Some(&filter.predicate)),
             LogicalPlan::CrossJoin(cross_join) => {
-                let predicates = split_conjunction_owned(filter.predicate.clone());
+                let predicates = split_conjunction_owned(filter.predicate);
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:832]: likewise here we avoid an additional clone

File: datafusion/optimizer/src/push_down_filter.rs [hunk 3]
```diff
@@ -861,16 +861,12 @@ impl OptimizerRule for PushDownFilter {
                     .collect();
                 let new_predicate: Vec<Expr> = zip
                     .filter(|(_, res)| res != &TableProviderFilterPushDown::Exact)
-                    .map(|(pred, _)| (*pred).clone())
+                    .map(|(pred, _)| pred.clone())
                     .collect();
 
                 let new_scan = LogicalPlan::TableScan(TableScan {
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:867]: 👍  this also looks like it might save some copying

**Review by alamb — APPROVED** [review]
> please help me to understand how the approach in the PR differs from what we have now. Its replaces clone on Arcs to Arc::clone, which imho the same https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references ?

I agree with @comphead  calling `Arc::clone` is the same thing, though it is more explicit that only an `Arc` is being copied so I think it is an improvement in the code

I think I found a few other Expr::clone  that this PR avoids, so from that perspective I think it is an improvement.

Thank you @comphead  and @erratic-pattern

**Comment by alamb** [comment]
Thanks again @erratic-pattern

**Comment by erratic-pattern** [comment]
Yes the `Arc::clone` is not a performance improvement, but was just a way for me to keep track of which clones were "zero copy" while working on this. It is a recommended convention as well, so I think it's worth keeping.

**Pull request merged**